### PR TITLE
Hw3

### PR DESCRIPTION
**Pull request by adlerberg0**

### CODE REVIEW

File: hw3/task1/hw1.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from typing import Any, Callable, Sequence
+
+
+def cache(times: int) -> Callable:
+    # use memoization technique
+    def cache_dec(func: Callable):
+        memoization = {}
+
+        def wrapper(*args: Sequence[Any]) -> Any:
```
Comment (AlekseiAfanasev) [inline hw3/task1/hw1.py:9]: `args` here would always be an empty tuple. So, memoization dict will be always like `{(): times_left}`

File: hw3/task1/hw1.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+            if args in memoization and memoization[args][1] > 0:
+                memoization[args][1] -= 1
+                return memoization[args][0]
+            else:
```
Comment (AlekseiAfanasev) [inline hw3/task1/hw1.py:15]: How can we make the decorator more simple here?

File: hw3/task2/hw2.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+def slow_calculate(value: Any):
+    """Some weird voodoo magic calculations"""
+    sleep(random.randint(1, 3))
+    data = hashlib.md5(str(value).encode()).digest()
+    return sum(struct.unpack("<" + "B" * len(data), data))
+
+
+def call_func(data: Sequence[Any], func: Callable = slow_calculate) -> Sequence[Any]:
+    p = Pool(25)
```
Comment (AlekseiAfanasev) [inline hw3/task2/hw2.py]: Please, read the warning block for this class: https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool

File: hw3/task3/hw3.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    def __init__(self, functions):
+        self.functions = functions
+
+    def apply(self, data):
+        return [item for item in data if all(i(item) for i in self.functions)]
+
+
+# example of usage:
+# positive_even = Filter(lambda a: a % 2 == 0, lambda a: a > 0, lambda a: isinstance(int, a)))
```
Comment (AlekseiAfanasev) [inline hw3/task3/hw3.py]: is this line correct?

File: hw3/task3/hw3.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+# positive_even.apply(range(100)) should return only even numbers from 0 to 99
+
+
+def make_filter(**keywords):
```
Comment (AlekseiAfanasev) [inline hw3/task3/hw3.py]: Could you please add typings for the code?

File: tests/hw3/test_task1.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+from io import StringIO
+from typing import Callable
+
+import pytest
+
+from hw3.task1.hw1 import cache
+
+
+def f():
```
Comment (AlekseiAfanasev) [inline tests/hw3/test_task1.py:9]: Why not just import the function from the completed task file?

File: tests/hw3/test_task1.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    return input("? ")  # careful with input() in python2, use raw_input() instead
+
+
+@pytest.mark.parametrize("func, times, repeat_pls", [(f, 1, 3), (f, 10, 2), (f, 30, 1)])
+def test_cache(monkeypatch, capsys, func: Callable, times: int, repeat_pls: int):
+    decorated = cache(times)(func)
+    test_str = f"{times}"
+    number_inputs = StringIO(test_str)
+    monkeypatch.setattr("sys.stdin", number_inputs)
+    for i in range((times + 1) * repeat_pls):
```
Comment (AlekseiAfanasev) [inline tests/hw3/test_task1.py:19]: Why `times + 1`?

File: tests/hw3/test_task1.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        number_inputs.seek(0)
+        if i % (times + 1) == 0:
+            number_inputs.seek(0)
+            res = decorated()
+            # resume capture after question have been asked
+            captured = capsys.readouterr()
+            assert captured.out == "? " and res == test_str
```
Comment (AlekseiAfanasev) [inline tests/hw3/test_task1.py]: Style improvement suggestion: if you add linebreaks before each `assert` statement, test file would become more readable

File: tests/hw3/test_task3.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        "last_name": "Gilbert",
+        "occupation": "was here",
+        "type": "person",
+    },
+    {
+        "name": "Smudge",
+        "type": "the Cat",
+        "preferences": "he no like vegetals",
+    },
+    {"is_dead": True, "kind": "parrot", "type": "bird", "name": "polly"},
```
Comment (AlekseiAfanasev) [inline tests/hw3/test_task3.py]: Reformat this list element please

File: tests/hw3/test_task1.py (hunk 8 above)
Comment (AlekseiAfanasev) [inline tests/hw3/test_task1.py]: And it's better to make assert with only one condition (break this into 2 asserts)

File: tests/hw3/test_task3.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    "params, data, expected_result ",
+    [
+        ({"name": "404", "type": "error"}, sample_data, []),
+        ({"name": "Bill", "occupation": "was here"}, sample_data, [sample_data[0]]),
+        ({"name": "Smudge", "type": "the Cat"}, sample_data, [sample_data[1]]),
+        ({"name": "polly", "type": "bird"}, sample_data, [sample_data[2]]),
+    ],
+)
+def test_cache(params: Dict[Any, Any], data: List[Any], expected_result: List[Any]):
+    assert expected_result == make_filter(**params).apply(data)
```
Comment (AlekseiAfanasev) [inline tests/hw3/test_task3.py]: Here and in other tests - it's kind of a naming convention - in pytest order should be actual == expected in asserts

Comment (AlekseiAfanasev) [reply tests/hw3/test_task3.py]: also, forgot to change name

File: tests/hw3/test_task4.py [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import pytest
+
+from hw3.task4.hw4 import is_armstrong
+
+
+@pytest.mark.parametrize("value, result", [(153, True), (10, False)])
+def test_cache(value: int, result: bool):
```
Comment (AlekseiAfanasev) [inline tests/hw3/test_task4.py]: For this function it is ok to use parametrize, but usually such cases should be splitted to two different test functions and be named like
```
def test_is_armstrong_{case_Description}_should_return_true():
   ...
# or simple 
def test_is_armstrong_{case_Description}_true():
   ...
# or simple 
def test_is_armstrong_{case_Description}_should_return_false():
   ...
# or simple 
def test_is_armstrong_{case_Description}_false():
   ...
```
also, forgot to change test function name

File: hw3/task4/hw4.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+### Example function signature and call
+"""
+
+
+def is_armstrong(number: int) -> bool:
+    chars_list = str(number)
+    digits_len = len(chars_list)
+    digits_list = list(map(lambda s: int(s) ** digits_len, chars_list))
+    return sum(digits_list) == number
```
Comment (AlekseiAfanasev) [inline hw3/task4/hw4.py:24]: Like adding line breaks before asserts, it is also often may be a good style to add lb before return statements, as it improves the code readability

File: hw3/task4/hw4.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+ - use map or other utilities from functools library,
+ - use anonymous functions (or use function as argument)
+ - do not use loops, preferably using list comprehensions
+
+### Example function signature and call
+"""
+
+
+def is_armstrong(number: int) -> bool:
+    chars_list = str(number)
```
Comment (AlekseiAfanasev) [inline hw3/task4/hw4.py]: How can we remove this variable?

File: tests/hw3/test_task3.py (hunk 10 above)
Comment (adlerberg0) [reply tests/hw3/test_task3.py]: done

File: tests/hw3/test_task4.py (hunk 11 above)
Comment (adlerberg0) [reply tests/hw3/test_task4.py]: done

File: tests/hw3/test_task1.py (hunk 8 above)
Comment (adlerberg0) [reply tests/hw3/test_task1.py]: done

File: hw3/task3/hw3.py (hunk 5 above)
Comment (adlerberg0) [reply hw3/task3/hw3.py]: Done

File: hw3/task3/hw3.py (hunk 4 above)
Comment (adlerberg0) [reply hw3/task3/hw3.py]: done

File: hw3/task4/hw4.py (hunk 13 above)
Comment (adlerberg0) [reply hw3/task4/hw4.py]: done

File: hw3/task4/hw4.py (hunk 12 above)
Comment (adlerberg0) [reply hw3/task4/hw4.py:24]: done

File: hw3/task2/hw2.py (hunk 3 above)
Comment (adlerberg0) [reply hw3/task2/hw2.py]: done

File: tests/hw3/test_task1.py (hunk 6 above)
Comment (adlerberg0) [reply tests/hw3/test_task1.py:9]: done

File: tests/hw3/test_task3.py (hunk 9 above)
Comment (adlerberg0) [reply tests/hw3/test_task3.py]: done

File: hw3/task1/hw1.py (hunk 1 above)
Comment (adlerberg0) [reply hw3/task1/hw1.py:9]: done

File: hw3/task1/hw1.py (hunk 2 above)
Comment (adlerberg0) [reply hw3/task1/hw1.py:15]: done

File: tests/hw3/test_task1.py (hunk 8 above)
Comment (adlerberg0) [reply tests/hw3/test_task1.py]: done

**Review by AlekseiAfanasev — APPROVED** [review]

**Pull request merged**